File: src/common/platform/intake.cpp
#include "src/common/platform/intake.hpp"

#include <algorithm>
#include <cstdio>
#include <cstdlib>
#include <filesystem>
#include <map>
#include <utility>

#include "src/common/platform/desktop.hpp"
#include "src/common/platform/inflate.hpp"
#include "src/common/platform/platform.hpp"
#include "src/common/platform/sha256.hpp"

namespace fs = std::filesystem;

namespace platform {
namespace intake {

namespace {

const int kUnpackerVersion = 1;

#ifdef _WIN32
const char kSep = '\\';
#else
const char kSep = '/';
#endif

bool slurp(const std::string &path, std::vector<uint8_t> *out) {
    std::FILE *f = std::fopen(path.c_str(), "rb");
    if (f == nullptr) return false;
    std::fseek(f, 0, SEEK_END);
    long n = std::ftell(f);
    std::fseek(f, 0, SEEK_SET);
    out->resize((size_t)(n < 0 ? 0 : n));
    if (!out->empty()) {
        size_t got = std::fread(out->data(), 1, out->size(), f);
        out->resize(got);
    }
    std::fclose(f);
    return true;
}

bool spill(const std::string &path, const std::vector<uint8_t> &bytes) {
    std::error_code ec;
    fs::create_directories(fs::path(path).parent_path(), ec);
    std::FILE *f = std::fopen(path.c_str(), "wb");
    if (f == nullptr) return false;
    if (!bytes.empty()) std::fwrite(bytes.data(), 1, bytes.size(), f);
    std::fclose(f);
    return true;
}

std::string stampField(const std::string &stamp, const char *key) {
    std::string needle = std::string("\"") + key + "\"";
    size_t at = stamp.find(needle);
    if (at == std::string::npos) return "";
    at = stamp.find(':', at + needle.size());
    if (at == std::string::npos) return "";
    size_t open = stamp.find('"', at);
    if (open == std::string::npos) return "";
    size_t close = stamp.find('"', open + 1);
    if (close == std::string::npos) return "";
    return stamp.substr(open + 1, close - open - 1);
}

std::string lower(std::string s) {
    for (size_t i = 0; i < s.size(); ++i) {
        if (s[i] >= 'A' && s[i] <= 'Z') s[i] = (char)(s[i] - 'A' + 'a');
    }
    return s;
}

bool endsWith(const std::string &s, const char *suffix) {
    size_t n = std::string(suffix).size();
    return s.size() >= n && s.compare(s.size() - n, n, suffix) == 0;
}

Variant byMembers(const std::vector<std::string> &names) {
    bool hasDatfiles = false;
    bool hasImgfiles = false;
    bool looseDat = false;
    bool hasCus = false;
    for (size_t i = 0; i < names.size(); ++i) {
        std::string n = lower(names[i]);
        if (n == "datfiles.lmp") hasDatfiles = true;
        if (n == "imgfiles.lmp") hasImgfiles = true;
        if (endsWith(n, ".dat")) looseDat = true;
        if (endsWith(n, ".cus")) hasCus = true;
    }
    bool hasLmp = hasDatfiles && hasImgfiles;
    if (hasLmp && !hasCus) return Variant::Dawnstar;
    if (hasCus && looseDat && !hasLmp) return Variant::Stormhold;
    return Variant::Unknown;
}

Variant byManifest(const std::string &text) {
    std::string t = lower(text);
    if (t.find("dawnstar") != std::string::npos) return Variant::Dawnstar;
    if (t.find("stormhold") != std::string::npos) return Variant::Stormhold;
    return Variant::Unknown;
}

std::string envOrEmpty(const char *name) {
#ifdef _WIN32
    char *value = nullptr;
    size_t len = 0;
    if (_dupenv_s(&value, &len, name) != 0 || value == nullptr) return std::string();
    std::string out(value);
    std::free(value);
    return out;
#else
    const char *value = std::getenv(name);
    return value == nullptr ? std::string() : std::string(value);
#endif
}

}

const char *variantId(Variant v) {
    switch (v) {
        case Variant::Dawnstar: return "dawnstar";
        case Variant::Stormhold: return "stormhold";
        default: return "unknown";
    }
}

Variant variantFromId(const std::string &id) {
    std::string v = lower(id);
    if (v == "dawnstar") return Variant::Dawnstar;
    if (v == "stormhold") return Variant::Stormhold;
    return Variant::Unknown;
}

namespace {
std::string &overrideDir() {
    static std::string dir;
    return dir;
}
}

void setUserDataDirForTesting(const std::string &dir) { overrideDir() = dir; }

std::string userDataDir() {
    if (!overrideDir().empty()) return overrideDir();
    std::string base = envOrEmpty("LOCALAPPDATA");
    if (base.empty()) base = envOrEmpty("XDG_DATA_HOME");
    if (base.empty()) {
        std::string home = envOrEmpty("HOME");
        if (!home.empty()) base = home + "/.local/share";
    }
    if (base.empty()) base = ".";
    std::string out = base + "/Raingrasp";
    for (size_t i = 0; i < out.size(); ++i) {
        if (out[i] == '/') out[i] = kSep;
    }
    return out;
}

void configurePortableMode(const std::string &exeDir, bool requested) {
    if (exeDir.empty()) return;
    std::error_code ec;
    const fs::path marker = fs::path(exeDir) / "portable.txt";
    if (requested || fs::is_regular_file(marker, ec)) {
        overrideDir() = (fs::path(exeDir) / "data").string();
    }
}

std::string treeDir(Variant v) {
    return userDataDir() + kSep + variantId(v);
}

std::string settingsDir() {
    return userDataDir() + kSep + "settings";
}

std::string saveDir(Variant v) {
    return userDataDir() + kSep + "saves" + kSep + variantId(v);
}

Variant identifyJar(const std::string &jarPath) {
    std::vector<uint8_t> bytes;
    if (!slurp(jarPath, &bytes)) return Variant::Unknown;

    std::vector<ZipEntry> entries;
    std::string error;
    if (!readZip(bytes, &entries, &error)) return Variant::Unknown;

    std::vector<std::string> names;
    std::string manifest;
    for (size_t i = 0; i < entries.size(); ++i) {
        names.push_back(entries[i].name);
        if (lower(entries[i].name) == "meta-inf/manifest.mf") {
            manifest.assign((const char *)entries[i].data.data(),
                            entries[i].data.size());
        }
    }

    Variant members = byMembers(names);
    if (members == Variant::Unknown) return Variant::Unknown;

    Variant manifestSays = byManifest(manifest);
    if (manifestSays != Variant::Unknown && manifestSays != members) {
        return Variant::Unknown;
    }
    return members;
}

std::vector<std::string> jarsIn(const std::string &dir) {
    std::vector<std::string> out;
    std::error_code ec;
    for (const fs::directory_entry &e : fs::directory_iterator(dir, ec)) {
        if (!e.is_regular_file(ec)) continue;
        if (lower(e.path().extension().string()) == ".jar") {
            out.push_back(e.path().string());
        }
    }
    std::sort(out.begin(), out.end());
    return out;
}

bool treeIsUsable(const std::string &dir, Variant v, std::string *why) {
    std::error_code ec;
    if (!fs::is_directory(dir, ec)) {
        if (why != nullptr) *why = "no such directory";
        return false;
    }

    std::vector<uint8_t> stampBytes;
    bool haveStamp = slurp(dir + "/stamp.json", &stampBytes);
    std::string stamp((const char *)stampBytes.data(), stampBytes.size());

    if (haveStamp && stamp.find("\"variant\"") != std::string::npos) {
        std::string want = std::string("\"") + variantId(v) + "\"";
        if (stamp.find(want) == std::string::npos) {
            if (why != nullptr) *why = "stamp.json records a different game";
            return false;
        }
    }

    if (v == Variant::Dawnstar) {
        static const char *required[] = {"datfiles.lmp", "imgfiles.lmp"};
        for (size_t i = 0; i < 2; ++i) {
            if (!fs::is_regular_file(dir + "/" + required[i], ec)) {
                if (why != nullptr) {
                    *why = std::string(required[i]) +
                           " is missing; the game reads it directly and cannot "
                           "start without it";
                }
                return false;
            }
        }
    } else if (v == Variant::Stormhold) {
        if (!fs::is_regular_file(dir + "/itemsin.dat", ec)) {
            if (why != nullptr) *why = "itemsin.dat is missing";
            return false;
        }
    }

    return true;
}

std::string treeContentDigest(const std::string &dir) {
    std::vector<uint8_t> bytes;
    if (!slurp(dir + "/stamp.json", &bytes)) return "";
    std::string stamp((const char *)bytes.data(), bytes.size());
    return stampField(stamp, "content_sha256");
}

bool unpackJar(const std::string &jarPath, const std::string &dest, Variant expect,
               std::string *error) {
    std::vector<uint8_t> bytes;
    if (!slurp(jarPath, &bytes)) {
        *error = "cannot read " + jarPath;
        return false;
    }

    std::vector<ZipEntry> entries;
    std::string zipError;
    if (!readZip(bytes, &entries, &zipError)) {
        *error = zipError;
        return false;
    }

    std::vector<std::string> names;
    std::string manifest;
    for (size_t i = 0; i < entries.size(); ++i) {
        names.push_back(entries[i].name);
        if (lower(entries[i].name) == "meta-inf/manifest.mf") {
            manifest.assign((const char *)entries[i].data.data(),
                            entries[i].data.size());
        }
    }

    Variant members = byMembers(names);
    if (members == Variant::Unknown) {
        *error =
            "that file does not look like either game. Dawnstar packs "
            "datfiles.lmp and imgfiles.lmp; Stormhold ships its .dat and .cus "
            "files loose.";
        return false;
    }
    Variant manifestSays = byManifest(manifest);
    if (manifestSays != Variant::Unknown && manifestSays != members) {
        *error = std::string("refusing that JAR: its members look like ") +
                 variantId(members) + " but its manifest says " +
                 variantId(manifestSays) + ".";
        return false;
    }
    if (expect != Variant::Unknown && expect != members) {
        *error = std::string("that is a ") + variantId(members) + " JAR, not " +
                 variantId(expect) + ".";
        return false;
    }

    std::vector<std::pair<std::string, const std::vector<uint8_t> *>> content;
    for (size_t i = 0; i < entries.size(); ++i) {
        content.push_back(std::make_pair(entries[i].name, &entries[i].data));
    }
    std::string digest = contentDigest(content);

    if (treeContentDigest(dest) == digest) {
        std::string why;
        if (treeIsUsable(dest, members, &why)) return true;
    }

    std::error_code ec;
    fs::create_directories(dest, ec);

    fs::path root = fs::weakly_canonical(fs::absolute(dest, ec), ec);
    size_t written = 0;
    std::string expanded;
    for (size_t i = 0; i < entries.size(); ++i) {
        fs::path target = fs::weakly_canonical(root / entries[i].name, ec);
        fs::path relative = target.lexically_relative(root);
        if (relative.empty() || relative.is_absolute() ||
            *relative.begin() == fs::path("..")) {
            *error = "refusing an entry outside the destination: " + entries[i].name;
            return false;
        }
        std::string t = target.string();
        if (!spill(t, entries[i].data)) {
            *error = "cannot write " + t;
            return false;
        }
        ++written;

        if (endsWith(lower(entries[i].name), ".lmp")) {
            std::vector<std::pair<std::string, std::vector<uint8_t>>> members2;
            if (parseLmpPreservingCase(entries[i].data, &members2)) {
                for (size_t m = 0; m < members2.size(); ++m) {
                    if (spill((root / members2[m].first).string(), members2[m].second)) {
                        ++written;
                    }
                }
                expanded += std::string(expanded.empty() ? "" : ", ") + "\"" +
                            entries[i].name + "\"";
            }
        }
    }

    if (written == 0) {
        *error = jarPath + " contained no files";
        return false;
    }

    char buf[1024];
    std::snprintf(buf, sizeof(buf),
                  "{\n  \"unpacker\": \"native\",\n  \"unpacker_version\": %d,\n"
                  "  \"variant\": \"%s\",\n  \"content_sha256\": \"%s\",\n"
                  "  \"files_written\": %u,\n"
                  "  \"archives_expanded\": [%s]\n}\n",
                  kUnpackerVersion, variantId(members), digest.c_str(),
                  (unsigned)written, expanded.c_str());
    std::string stamp(buf);
    spill((root / "stamp.json").string(),
          std::vector<uint8_t>(stamp.begin(), stamp.end()));
    return true;
}

Variant identifyTree(const std::string &dir) {
    std::string why;
    if (treeIsUsable(dir, Variant::Dawnstar, &why)) return Variant::Dawnstar;
    if (treeIsUsable(dir, Variant::Stormhold, &why)) return Variant::Stormhold;
    return Variant::Unknown;
}

Result resolve(const Request &request) {
    Result result;
    result.variant = request.want;

    if (!request.dataDir.empty()) {
        std::string why;
        Variant v = request.want == Variant::Unknown ? identifyTree(request.dataDir)
                                                     : request.want;
        if (v != Variant::Unknown && treeIsUsable(request.dataDir, v, &why)) {
            result.status = Result::Status::Ready;
            result.variant = v;
            result.tree = request.dataDir;
            return result;
        }
        if (v == Variant::Unknown) why = "it carries no intake stamp naming a game";
        result.status = Result::Status::Refused;
        result.message = request.dataDir + " is not a usable game directory: " + why;
        return result;
    }

    std::string jar = request.jarPath;
    if (jar.empty() && !request.bareArg.empty()) {
        std::error_code ec;
        if (fs::is_directory(request.bareArg, ec)) {
            std::string why;
            Variant v = request.want == Variant::Unknown ? identifyTree(request.bareArg)
                                                         : request.want;
            if (v != Variant::Unknown && treeIsUsable(request.bareArg, v, &why)) {
                result.status = Result::Status::Ready;
                result.variant = v;
                result.tree = request.bareArg;
                return result;
            }
            if (v == Variant::Unknown) why = "it carries no intake stamp naming a game";
            result.status = Result::Status::Refused;
            result.message =
                request.bareArg + " is not a usable game directory: " + why;
            return result;
        }
        jar = request.bareArg;
    }

    if (!jar.empty()) {
        Variant expect = request.want;
        if (expect == Variant::Unknown) {
            expect = identifyJar(jar);
            if (expect == Variant::Unknown) {
                result.status = Result::Status::Refused;
                result.message = jar + " is not a Dawnstar or Stormhold JAR";
                return result;
            }
        }
        std::string dest = treeDir(expect);
        std::string error;
        if (unpackJar(jar, dest, expect, &error)) {
            result.status = Result::Status::Ready;
            result.variant = expect;
            result.tree = dest;
            return result;
        }
        Variant actual = identifyJar(jar);
        if (actual != Variant::Unknown && actual != request.want) {
            std::string otherDest = treeDir(actual);
            std::string otherError;
            if (unpackJar(jar, otherDest, actual, &otherError)) {
                result.status = Result::Status::Refused;
                result.message = std::string("that is a ") + variantId(actual) +
                                 " JAR, not " + variantId(request.want) +
                                 ". It has been unpacked into " + otherDest +
                                 " for the " + variantId(actual) + " build.";
                return result;
            }
        }
        result.status = Result::Status::Refused;
        result.message = error;
        return result;
    }

    if (request.want == Variant::Unknown) {
        Result found[2];
        int hits = 0;
        const Variant both[] = {Variant::Dawnstar, Variant::Stormhold};
        for (Variant v : both) {
            Request narrowed = request;
            narrowed.want = v;
            Result r = resolve(narrowed);
            if (r.status == Result::Status::Ready) found[hits++] = r;
        }
        if (hits == 1) return found[0];
        if (hits == 2) {
            result.status = Result::Status::Refused;
            result.message =
                "Both games' data are present. Say which to play: "
                "--game dawnstar or --game stormhold.";
            return result;
        }
        result.status = Result::Status::NeedsJar;
        result.message =
            "No game data found.\n\n"
            "This port ships no game assets. Supply your own copy of the "
            "Dawnstar or Stormhold .jar in any of these ways:\n"
            "  - drag the .jar onto this program\n"
            "  - pass it:  --jar <path-to.jar>\n"
            "  - put it beside this program, or in " + userDataDir() + "\n"
            "  - point at an already-unpacked directory:  --data <dir>\n"
            "  - or unpack it beside this program as ./dawnstar or ./stormhold"
            "\n\n"
            "It is identified by its contents, so the filename does not matter.";
        return result;
    }

    std::string tree = treeDir(request.want);
    std::string why;
    if (treeIsUsable(tree, request.want, &why)) {
        result.status = Result::Status::Ready;
        result.tree = tree;
        return result;
    }

    if (!request.exeDir.empty()) {
        std::string beside = request.exeDir + kSep + variantId(request.want);
        std::string besideWhy;
        if (treeIsUsable(beside, request.want, &besideWhy)) {
            result.status = Result::Status::Ready;
            result.tree = beside;
            return result;
        }
    }

    std::vector<std::string> candidates;
    if (!request.exeDir.empty()) {
        std::vector<std::string> found = jarsIn(request.exeDir);
        candidates.insert(candidates.end(), found.begin(), found.end());
    }
    std::vector<std::string> inData = jarsIn(userDataDir());
    candidates.insert(candidates.end(), inData.begin(), inData.end());

    for (size_t i = 0; i < candidates.size(); ++i) {
        if (identifyJar(candidates[i]) != request.want) continue;
        std::string error;
        if (unpackJar(candidates[i], tree, request.want, &error)) {
            result.status = Result::Status::Ready;
            result.tree = tree;
            return result;
        }
    }

    result.status = Result::Status::NeedsJar;
    result.message =
        std::string("No ") + variantId(request.want) +
        " game data found.\n\n"
        "This port ships no game assets. Supply your own copy of the " +
        variantId(request.want) +
        " .jar in any of these ways:\n"
        "  - drag the .jar onto this program\n"
        "  - pass it:  --jar <path-to.jar>\n"
        "  - put it beside this program, or in " + userDataDir() + "\n"
        "  - point at an already-unpacked directory:  --data <dir>\n"
        "  - or unpack it beside this program as ./" + variantId(request.want) +
        "\n\n"
        "It is identified by its contents, so the filename does not matter.";
    return result;
}

}
}

File: bazel/intake.py
"""Identify a game JAR by content and unpack it into a variant tree.

This is Phase 3's intake. It answers two questions the runtime must not have to
ask: *which game is this*, and *where do its bytes live*.

**Identification is by content, never by filename.** A user's download is called
whatever their browser named it, so the member list is the only reliable signal:

    Dawnstar    packs its data into datfiles.lmp and imgfiles.lmp
    Stormhold   ships its .dat and .cus files loose

`META-INF/MANIFEST.MF` corroborates but cannot decide alone, and the reason is
worth stating: Dawnstar's `MIDlet-1` says `DawnStar`, but Stormhold's
`MIDlet-Name` is the series title `The Elder Scrolls` and names the game
nowhere. So members decide, the manifest cross-checks, and a disagreement is
refused rather than guessed -- unpacking a misidentified JAR into the wrong
variant tree would produce a game that boots and is subtly wrong, which is far
worse than a refusal.

**`.lmp` archives are expanded alongside, never instead of, themselves.** Once
`datfiles.lmp`'s members are loose in the tree, the archive stops being special
for *data*: the members are ordinary files in the variant layer, so
`override/helptext.dat` shadows one without anybody repacking an archive.

The archives are nonetheless always kept, and `imgfiles.lmp` is why. Expanding
it is useful -- it makes `override/panel.png` possible the same way -- but it
does **not** make the archive redundant, because `Game::createImageFromFile`
reads `imgfiles.lmp` itself to build an image table keyed by byte offset, and
`Game::createImage` serves every sprite out of that table rather than through
the resource path. Delete the archive and the table is empty, `createImage`
returns null, and the game dereferences it during startup.

So: expand for overridability, keep for correctness. A tree missing its `.lmp`
files is not a supported state, and `verify` reports it as damage rather than as
a tree that merely lacks a fallback.

**A stamp records what was unpacked.** `stamp.json` holds the source JAR's size
and hash plus this unpacker's version, which is what makes "did the user swap
JARs, is this tree stale" answerable without rescanning every file.

Usage:
    intake.py identify <archive.jar>
    intake.py unpack <archive.jar> <output-dir> [--expect <variant>]
    intake.py verify <tree-dir>
"""

import hashlib
import io
import json
import os
import struct
import sys
import zipfile

# Bumped when the unpacked layout changes in a way that makes an existing tree
# wrong rather than merely old. A tree whose stamp records a lower version needs
# re-unpacking; one that merely records a different JAR hash needs the user's
# attention instead.
UNPACKER_VERSION = 1

VARIANTS = ("dawnstar", "stormhold")


class Refused(Exception):
    """Identification failed, or the two signals disagreed."""


# --------------------------------------------------------------- the .lmp
#
# Layout, confirmed against the shipped archive rather than from the prose:
# a run of `'-' name '-' offset:be32 size:be16` directory entries packed at the
# start of the file, terminated by the first byte that is not `-`, then the data
# region. See docs/DATA_FORMATS.md, which described a different (and wrong)
# format until Phase 3.


def parse_lmp(data):
    """Returns [(name, bytes)] in directory order. Raises on a damaged archive."""
    members = []
    pos = 0
    while pos < len(data) and data[pos:pos + 1] == b"-":
        pos += 1
        end = data.find(b"-", pos)
        if end < 0:
            raise Refused("lmp: end of archive while reading a member name")
        name = data[pos:end].decode("ascii", "replace")
        pos = end + 1
        if pos + 6 > len(data):
            raise Refused("lmp: truncated directory entry for %s" % name)
        offset, size = struct.unpack(">IH", data[pos:pos + 6])
        pos += 6
        if offset + size > len(data):
            raise Refused(
                "lmp: member %s extends past the end of the archive" % name)
        members.append((name, data[offset:offset + size]))
    return members


# -------------------------------------------------------- identification


def _members(zf):
    return set(
        info.filename for info in zf.infolist() if not info.is_dir())


def _manifest(zf):
    for name in ("META-INF/MANIFEST.MF", "META-INF/manifest.mf"):
        try:
            return zf.read(name).decode("utf-8", "replace")
        except KeyError:
            continue
    return ""


def _by_members(names):
    lowered = set(n.lower() for n in names)
    has_lmp = "datfiles.lmp" in lowered and "imgfiles.lmp" in lowered
    # Stormhold's tell is loose data plus the .cus sprites Dawnstar has none of.
    loose_dat = any(n.endswith(".dat") for n in lowered)
    has_cus = any(n.endswith(".cus") for n in lowered)
    if has_lmp and not has_cus:
        return "dawnstar"
    if has_cus and loose_dat and not has_lmp:
        return "stormhold"
    return None


def _by_manifest(text):
    """The cross-check. Returns a variant, or None when the manifest is silent.

    Deliberately weaker than the member check. Stormhold's manifest names the
    series and not the game, so `None` here is the *expected* answer for a valid
    Stormhold JAR and must never be treated as a failure.
    """
    lowered = text.lower()
    if "dawnstar" in lowered:
        return "dawnstar"
    if "stormhold" in lowered:
        return "stormhold"
    return None


def identify(path):
    """Returns (variant, evidence). Raises Refused when it cannot be sure."""
    with zipfile.ZipFile(path) as zf:
        names = _members(zf)
        manifest = _manifest(zf)

    by_members = _by_members(names)
    by_manifest = _by_manifest(manifest)

    if by_members is None:
        raise Refused(
            "cannot identify %s: its members match neither game. Dawnstar packs "
            "datfiles.lmp and imgfiles.lmp; Stormhold ships .dat and .cus files "
            "loose." % os.path.basename(path))

    # A manifest that names the *other* game is the case worth refusing loudly:
    # it means one of the two signals is wrong, and guessing which would risk
    # unpacking into the wrong tree.
    if by_manifest is not None and by_manifest != by_members:
        raise Refused(
            "refusing %s: its members look like %s but its manifest says %s. "
            "One of the two is wrong, and guessing could unpack it into the "
            "wrong game's tree." % (os.path.basename(path), by_members, by_manifest))

    evidence = {
        "by_members": by_members,
        "by_manifest": by_manifest,
        "manifest_corroborates": by_manifest == by_members,
        "member_count": len(names),
    }
    return by_members, evidence


# --------------------------------------------------------------- unpacking


def _sha256(path):
    h = hashlib.sha256()
    with open(path, "rb") as handle:
        while True:
            chunk = handle.read(1 << 16)
            if not chunk:
                break
            h.update(chunk)
    return h.hexdigest()


def unpack(archive, out_dir, expect=None):
    """Unpacks `archive` into `out_dir`, expanding any .lmp it contains.

    Returns (variant, files_written). Raises Refused when identification fails
    or contradicts `expect` -- which is how a JAR dropped onto the wrong picker
    row is reported rather than unpacked.
    """
    variant, evidence = identify(archive)
    if expect is not None and expect != variant:
        raise Refused(
            "that is a %s JAR, not %s. Drop it on the %s row, or let intake "
            "unpack it into %s/ for the other binary."
            % (variant, expect, variant, variant))

    root = os.path.abspath(out_dir)
    if not os.path.isdir(root):
        os.makedirs(root)

    written = []
    archives_expanded = []

    def write(name, data):
        # Path traversal is an error rather than something to sanitise away:
        # these archives are hash-pinned, so a traversal entry would mean the
        # manifest is wrong, and that is worth failing loudly over.
        target = os.path.abspath(os.path.join(root, name))
        if not target.startswith(root + os.sep) and target != root:
            raise Refused("refusing entry outside the output directory: %s" % name)
        parent = os.path.dirname(target)
        if parent and not os.path.isdir(parent):
            os.makedirs(parent)
        with open(target, "wb") as dst:
            dst.write(data)
        written.append(name)

    content = hashlib.sha256()
    with zipfile.ZipFile(archive) as zf:
        # Name order, so two archives holding the same files agree however their
        # central directories happened to be laid out. The framing below --
        # name, NUL, big-endian length, member digest -- is shared with
        # `content_tree_digest` in bazel/jar_inventory.py and with
        # `contentDigest` in the native unpacker; all three must stay in step.
        for info in sorted(
            (e for e in zf.infolist() if not e.is_dir()),
            key=lambda e: e.filename,
        ):
            content.update(info.filename.encode("utf-8"))
            content.update(b"\0")
            content.update(len(zf.read(info)).to_bytes(8, byteorder="big"))
            content.update(hashlib.sha256(zf.read(info)).digest())

        for info in zf.infolist():
            if info.is_dir():
                continue
            data = zf.read(info)
            write(info.filename, data)
            # Expand the archive *as well as* keeping it. Keeping it costs a few
            # kilobytes and means a tree stays readable by the runtime's archive
            # layer; expanding it is what makes override/ work per member.
            if info.filename.lower().endswith(".lmp"):
                members = parse_lmp(data)
                for name, body in members:
                    write(name, body)
                archives_expanded.append(
                    {"archive": info.filename, "members": len(members)})

    if not written:
        raise Refused("%s contained no files" % archive)

    stamp = {
        "unpacker_version": UNPACKER_VERSION,
        "variant": variant,
        # Keyed on content, not on the archive: the two Stormhold JARs hold the
        # same files in different ZIP packaging, so the archive hash above would
        # call them different games' data and re-unpack on a swap. The native
        # unpacker writes this same field, computed the same way, so a tree from
        # either is legible to both.
        "content_sha256": content.hexdigest(),
        "source": {
            "name": os.path.basename(archive),
            "size": os.path.getsize(archive),
            "sha256": _sha256(archive),
        },
        "identification": evidence,
        "files_written": len(written),
        "archives_expanded": archives_expanded,
    }
    with io.open(os.path.join(root, "stamp.json"), "w", encoding="utf-8") as handle:
        handle.write(json.dumps(stamp, indent=2, sort_keys=True))
        handle.write("\n")

    return variant, written


def read_stamp(tree):
    """Returns the stamp dict for an unpacked tree, or None when absent."""
    path = os.path.join(tree, "stamp.json")
    if not os.path.isfile(path):
        return None
    try:
        with io.open(path, encoding="utf-8") as handle:
            return json.load(handle)
    except ValueError:
        return None


def is_stale(tree, archive=None):
    """Why `tree` needs re-unpacking, or None when it is current.

    This is the question the stamp exists to answer without rescanning: an
    unpacked tree from an older layout, or one produced from a different JAR
    than the user now has, is stale in two quite different ways and the caller
    wants to tell them apart.
    """
    stamp = read_stamp(tree)
    if stamp is None:
        return "no stamp.json: the tree was not produced by intake"
    if stamp.get("unpacker_version", 0) < UNPACKER_VERSION:
        return "unpacked by an older intake (version %s, current %s)" % (
            stamp.get("unpacker_version"), UNPACKER_VERSION)
    if archive is not None:
        want = _sha256(archive)
        got = stamp.get("source", {}).get("sha256")
        if want != got:
            return "unpacked from a different JAR (tree %s, supplied %s)" % (
                (got or "?")[:12], want[:12])
    return None


def verify(tree):
    """Returns a list of problems with an unpacked tree; empty means intact.

    Distinct from `is_stale`, which asks whether the tree matches the JAR the
    user now has. This asks whether the tree is *usable at all*, and the case it
    exists for is a missing `.lmp`: `Game::createImageFromFile` reads
    `imgfiles.lmp` directly to build its sprite table, so a Dawnstar tree
    without it segfaults during startup rather than failing to find a file.
    That is a hard crash a long way from its cause, so it is worth naming here.
    """
    stamp = read_stamp(tree)
    if stamp is None:
        return ["no stamp.json: the tree was not produced by intake"]

    problems = []
    for entry in stamp.get("archives_expanded", []):
        name = entry.get("archive")
        if name and not os.path.isfile(os.path.join(tree, name)):
            problems.append(
                "%s is missing. Its members were expanded into the tree, but the "
                "archive itself is still required: the game reads it directly to "
                "build its image table, and without it startup crashes." % name)

    if not stamp.get("files_written"):
        problems.append("stamp.json records no files_written")

    return problems


# ------------------------------------------------------------------- cli


def main(argv):
    if len(argv) >= 3 and argv[1] == "identify":
        try:
            variant, evidence = identify(argv[2])
        except Refused as exc:
            sys.stderr.write("%s\n" % exc)
            return 1
        sys.stdout.write("%s\n" % variant)
        sys.stdout.write("  members say:  %s\n" % evidence["by_members"])
        sys.stdout.write("  manifest says: %s\n" % (
            evidence["by_manifest"] or "nothing (expected for Stormhold)"))
        return 0

    if len(argv) >= 4 and argv[1] == "unpack":
        expect = None
        if "--expect" in argv:
            expect = argv[argv.index("--expect") + 1]
        try:
            variant, written = unpack(argv[2], argv[3], expect)
        except Refused as exc:
            sys.stderr.write("%s\n" % exc)
            return 1
        sys.stdout.write("unpacked %s: %d files into %s\n"
                         % (variant, len(written), argv[3]))
        return 0

    if len(argv) >= 3 and argv[1] == "verify":
        problems = verify(argv[2])
        if problems:
            for line in problems:
                sys.stderr.write("%s\n" % line)
            return 1
        sys.stdout.write("%s: intact\n" % argv[2])
        return 0

    sys.stderr.write(
        "usage: intake.py identify <archive.jar>\n"
        "       intake.py unpack <archive.jar> <output-dir> [--expect <variant>]\n"
        "       intake.py verify <tree-dir>\n")
    return 2


if __name__ == "__main__":
    sys.exit(main(sys.argv))
